### PR TITLE
Improve support for syntax highlighting

### PR DESCRIPTION
**Pull request by gcluzel**
This PR adds features requested in #412. It may introduce conflicts with PR #411.

* Resolve bugs in syntax highlighting
* Adding procedure calls

### CODE REVIEW

**Comment by Entomy** [comment]
Since you're concerned about possible conflicts, do you want to wait on reviewing this until #411 is committed and you can rebase?

**Comment by gcluzel** [comment]
You can review it without problem. I added the `basic_declarative_item` construction in my both PR because it's more clear when I have it. I can rebase and amend my commit after if needed

File: integration/vscode/ada/syntaxes/ada.tmLanguage.json [hunk 1]
```diff
@@ -183,13 +183,26 @@
 				"7": { "patterns": [ { "include": "#exponent_part" } ] }
 			}
 		},
+		"declarative_item": {
+			"patterns": [
+				{ "include": "#body" },
+				{ "include": "#basic_declarative_item" }
+			]
+		},
```
Comment (gcluzel) [inline integration/vscode/ada/syntaxes/ada.tmLanguage.json:191]: This part doesn't work. Whatever the way we put it (`body` first or `basic_declarative_item` first), either `function_body` or `function_specification` is matched.
Maybe the solution is to merge the two definitions. @Entomy you may have better solution.

Comment (Entomy) [reply integration/vscode/ada/syntaxes/ada.tmLanguage.json:191]: I just got back to where I have my own computer. I'll look at this in about an hour.

**Comment by Entomy** [comment]
@gcluzel will you rebase this? There's a regression and I'm not entirely sure why. That's the easiest first thing to check.

**Comment by gcluzel** [comment]
Sorry, I forgot to push

**Comment by Entomy** [comment]
That was it. Looks good to me!

**Comment by gcluzel** [comment]
Good! But can you just check this snippet
```Ada
with Ada.Text_IO;

procedure Main is
   package B is
      procedure C;
   end B;

   package body B is
      procedure C is
      begin
         null;
      end C;
   end B;

   procedure D
   with
      Import => True;
   
   procedure E is
   begin
      Ada.Text_IO.Put_Line("Hey");
   end E;
begin
end Main;
```
It should fail

**Comment by Entomy** [comment]
🤦🏻‍♂️ oh yeah. I forgot this was related to #412. 

image

This is what I'm seeing. `procedure E` makes sense to fail, because that's not a valid construct, and will have to be handled later on by the failover classifier. `procedure D` on the other hand, is incorrectly classified as a procedure body (`meta.declaration.procedure.body.ada`). Since it breaks beyond this point, this misclassification is almost certainly what's responsible, as that pattern will be looking for an `end D;` which doesn't exist. This is happening because the order of elements within a `patterns` array is important, and you've put matching a procedure body in front of matching a procedure specification. Because the spec hasn't been consumed, and a spec will match the `begin` part of a body, it thinks there's a procedure body there, and handles everything as if there is.

**Comment by gcluzel** [comment]
But `procedure E` is a valid construction in Ada and we have to handle it. In the [Ada RM](https://www.adaic.org/resources/add_content/standards/05rm/html/RM-6-3.html), this construction is allowed and not rejected by gcc.
Then both `procedure D` and `procedure E` have to be handled.

> This is happening because the order of elements within a patterns array is important, and you've put matching a procedure body in front of matching a procedure specification. Because the spec hasn't been consumed, and a spec will match the begin part of a body, it thinks there's a procedure body there, and handles everything as if there is.

I agree with this. This is why I put the `body` scope first, otherwise `package body` is handled as a package declaration.
But this doesn't work for procedures because both constructions exist and both constructions start with the same pattern (which is `(?i)\\b(procedure)\\s+((?:\\w|\\d|\\.|_)+)\\b`). We cannot distinguish which one is supposed to be matched as you mentioned (and even if we put `basic_declarative_item` in first our problem is not solved). So maybe it's necessary to merge both pattern.

**Comment by Entomy** [comment]
> But procedure E is a...

Yeah I was thinking about in a different context. I need to finish my coffee before doing these reviews.

I'll have to think this over, and play with things. I'll get back to you soon on this.

**Comment by gcluzel** [comment]
> I'll get back to you soon on this.

Thank you for your help! And do not hesitate to add your own commits.

**Comment by Entomy** [comment]
Here's an idea:
Put the specs before the bodies, use `is` as the discriminator. That would mean `is` has to be in the `begin` part, but it removes any ambiguity. And all of ARM, GNAT, and my own style have `is` on the same line, so this shouldn't be an issue in practice.

**Comment by Entomy** [comment]
I said that backwards. Using `is` as the discriminator for subroutine bodies requires putting the bodies _before_ the specs, otherwise they'll all be classified as specs.

**Comment by gcluzel** [comment]
I have some reservations about doing that. We also have to match the parameters of the procedure/function. Then a lot of programs won't be match, for example:
```Ada
procedure Add_Descriptor
   (Process : Process_Access;
   Kind    : Pipe_Kinds;
   fd      : Interfaces.C.int)
is
```
But maybe I have another idea. Let's play!

**Comment by Entomy** [comment]
> Then a lot of programs won't be match...

No, you're right, that's an excellent point. Hmm.

> But maybe I have another idea. Let's play!

Oh?

**Comment by gcluzel** [comment]
> Oh?

Finally no. That was a bad idea. I don't see anything else than matching for `procedure_body`
```JSON
"procedure_body": {
	"name": "meta.declaration.procedure.body.ada",
	"begin": "(?i)\\b(overriding\\s+)?(procedure)\\s+((?:\\w|\\d|\\.|_)+)\\b",
	"end": "(?i)(?:\\b(end)\\s+(\\2)\\s*)?(;)",
	....
```
That is very imprecise because we also match incorrect programs. But at least we highlight correctly all correct programs.

**Comment by setton** [comment]
> That is very imprecise because we also match incorrect programs. But at least we highlight correctly all correct programs.

Note that in general it's not a problem to match incorrect programs... as @raph-amiard pointed out, it's something we want. When you are typing things in a source editor, 99% of the time the code is invalid Ada, and we still want the highlighter to do a "good enough guess" highlighting of what's there...

**Comment by Entomy** [comment]
And that's the intention anyways. Much like I said in the issue he raised, there is a fallback mechanism I've always accounted for, right down to exactly how to add it, as well as an explanation of why it's not yet there.

The specific and highly structured nature of the approach taken is to handle context within the language, for an overall more detailed highlighting experience. It's a lot of work just to get some distinction between certain keywords, or actually being able to highlight types in the majority of cases. That being said, like I mentioned there as well, the classifier is not so specific and accurate as to attempt to be a "conforming processor" and certain things have been made more generalized in an effort to keep the classifier working even in failure cases. That may be an option here, where is-begin-end is handled as part of a subroutine declaration internally, rather than a distinct declaration.

The most ideal approach is something I'm not sure how to implement myself, and I'm not sure if you guys are ready to yet. There is a mechanism for an LSP to actually provide classifications of tokens. This is utilized by the TreeSitter based extensions and is the most ideal way to accomplish many of these things, with the tmLanguage files providing quick and easy stuff. To see why, consider the following:

~~~~ ada
declare
    A : Integer := Integer(42);
begin
    -- whatever
end;
~~~~

There is no way, based solely on grammar, to correctly determine `Integer(42)` is a cast and could be classified as a type, and not a subroutine call.

**Comment by setton** [comment]
Given the sophistication of the current engine, I think the first order of priority is to find a way to automate testing for it before we add more sophistication... I'm trying to find the time to have a go at it myself.

**Comment by gcluzel** [comment]
I'm pushing my solution in a separate commit. It's maybe not the best solution, but I don't see any other until a semantic highlighter is implemented. Feel free to give me your feeling about it.

**Comment by Entomy** [comment]
@setton absolutely. The current approach is still correct in the overwhelming majority of cases, and just needs to be developed to completeness. Jumping straight to another design so quickly is foolish.

@gcluzel I won't be able to look until about 21:00 EST, but I will then.

**Comment by setton** [comment]
On my side I started a scaffolding for setting up automated tests. Now I need to learn the API to add real tests in it...

**Comment by Entomy** [comment]
@setton I saw. I'll try to familiarize myself as well and help you out there. With everything else going on I'm not so sure how much time I can devote to this, but I can chip in a bit.

@gcluzel I'm now at the hospital until 24:00 not 21:00, and I've been up since 6:00 so I'll review tomorrow.

**Comment by gcluzel** [comment]
@Entomy Don't worry, there is nothing urgent at all. Take care of you

**Comment by Entomy** [comment]
Oh I'm fine. Just working. Taking care of others.

**Comment by Entomy** [comment]
Looks good, actually. What you did seems to have worked.

On a maybe related note, `goto` and `exit` regressed at some point but I don't know if that happened with this one or elsewhere. Also, I've been noticing the `repository` increasingly getting out of alphabetical order; please keep it that way as it makes scanning through things easier (we can consider another organizational system, that's just what I went with for now.

**Comment by gcluzel** [comment]
> Also, I've been noticing the repository increasingly getting out of alphabetical order; please keep it that way as it makes scanning through things easier (we can consider another organizational system, that's just what I went with for now.

Thanks for this precision. I wondered how was organised the item. Now I know. I'll change that. And see for `goto` and `exit`.
(And also sorry for the deletion of the branch)

**Comment by Entomy** [comment]
Hey man, it happens, no worries. I didn't have any documentation for these things because typically speaking, you just don't see many contributors in FOSS projects; I'll add this to the `CONTRIBUTING.md` file. Realistically speaking most FOSS is developed by 1-5 people, even for absolutely massive projects. Your contributions are ***very*** appreciated, especially as I'm unable to regularly contribute a lot between work and issues that developed making it difficult to use VS Code itself. *Thank you*

**Comment by gcluzel** [comment]
> On a maybe related note, `goto` and `exit` regressed at some point but I don't know if that happened with this one or elsewhere.

For `goto`, I don't know regression you saw. But in commit e4336d8 (line 620) you delete the `goto` patterns.
For `exit`, I guess that it comes from the same commit, line 450. You now match `#label` which is `"(<<)?((?:\\w|\\d|_)+)(:|>>)"`. Maybe you might add a `?` for the last group, or change for what it was before.

**Comment by gcluzel** [comment]
- I merged my two previous commits and ordered everything in alphabetic order.
- I resolve the when and exit regression in a separate commit.

**Comment by Entomy** [comment]
> For goto, I don't know regression you saw. But in commit e4336d8 (line 620) you delete the goto patterns.

🤦🏻‍♂️

> For **exit**...

Yeah, I'll look at this one myself after this PR is committed.

> Maybe you might add a ? for the last group, or change for what it was before.

Yeah, this was because Ada technically has two different label types, which is... not a decision I agree with. I'll figure out how to handle this well.

**Comment by gcluzel** [comment]
> Yeah, I'll look at this one myself after this PR is committed.

Ok! I revert my last commit and you'll see ;)

**Comment by setton** [comment]
Let's approve/merge this and build on top of it. Thanks a lot for the contribution and the review @Entomy and @gcluzel. If one of you could approve #422, I'll then take care of adding test cases to cover this!

**Pull request merged**